File: src/main/java/dark/leech/text/plugin/js/api/JsScriptEngine.java
package dark.leech.text.plugin.js.api;

import java.util.Map;

import org.mozilla.javascript.Context;
import org.mozilla.javascript.NativeArray;
import org.mozilla.javascript.Scriptable;

import dark.leech.text.action.Log;

/**
 * JavaScript script engine with resource management and context pooling. Uses Rhino engine for
 * vBook plugin compatibility.
 */
public class JsScriptEngine {

    private static final Object LOCK = new Object();
    private static JsScriptEngine engine;
    private final RhinoContextPool contextPool;

    // Resource limits
    private static final int MAX_CONTEXTS = 5;
    private static final int MAX_EXECUTION_TIME_MS = 30_000; // 30 seconds
    private static final int MAX_MEMORY_MB = 50;

    private JsScriptEngine() {
        this.contextPool = new RhinoContextPool(MAX_CONTEXTS, MAX_EXECUTION_TIME_MS, MAX_MEMORY_MB);
        Log.add("Rhino JavaScript engine initialized");
    }

    /**
     * Gets the singleton instance of JsScriptEngine.
     *
     * @return The JsScriptEngine instance
     */
    public static JsScriptEngine getInstance() {
        if (engine == null) {
            synchronized (LOCK) {
                if (engine == null) {
                    engine = new JsScriptEngine();
                }
            }
        }
        return engine;
    }

    /** Resets the singleton instance. Used primarily for testing. */
    public static void resetInstance() {
        synchronized (LOCK) {
            engine = null;
        }
    }

    /**
     * Executes a JavaScript script with resource limits and timeout.
     *
     * @param script The JavaScript code to execute
     * @return The result of the script execution
     * @throws ExecutionException if execution fails or times out
     */
    public Object execute(String script)
            throws ExecutionException, RhinoContextPool.ResourceExhaustedException {
        RhinoPooledContext pooledContext = null;
        Context ctx = null;
        try {
            // Borrow from pool (creates if needed)
            pooledContext = contextPool.borrowContext();
            ctx = pooledContext.getContext();
            ctx.setLanguageVersion(200); // ES6 support for vBook compatibility
            Scriptable scope = ctx.initStandardObjects();

            // Execute script
            Object result = ctx.evaluateString(scope, script, "script", 1, null);
            return result;

        } catch (Exception e) {
            if (pooledContext != null) {
                contextPool.invalidateContext(pooledContext);
                pooledContext = null;
            }
            throw new ExecutionException("Script execution failed: " + e.getMessage(), e);
        } finally {
            if (pooledContext != null) {
                contextPool.returnContext(pooledContext);
            } else if (ctx != null) {
                Context.exit();
            }
        }
    }

    /**
     * Executes a JavaScript function with parameters.
     *
     * @param script The script containing the function
     * @param functionName The name of the function to call
     * @param args Arguments to pass to the function
     * @return The result of the function call
     * @throws ExecutionException if execution fails
     */
    public Object executeFunction(String script, String functionName, Object... args)
            throws ExecutionException, RhinoContextPool.ResourceExhaustedException {
        RhinoPooledContext pooledContext = null;
        Context ctx = null;
        try {
            pooledContext = contextPool.borrowContext();
            ctx = pooledContext.getContext();
            ctx.setLanguageVersion(200); // ES6 support for vBook compatibility
            Scriptable scope = ctx.initStandardObjects();

            // Evaluate the script first
            ctx.evaluateString(scope, script, "script", 1, null);

            // Get the function from scope
            Object functionObj = scope.get(functionName, scope);
            if (!(functionObj instanceof org.mozilla.javascript.Function)) {
                throw new ExecutionException(
                        "Function '" + functionName + "' not found or not executable");
            }

            // Execute the function with arguments
            org.mozilla.javascript.Function function =
                    (org.mozilla.javascript.Function) functionObj;
            Object result = function.call(ctx, scope, scope, args);
            return result;

        } catch (ExecutionException e) {
            throw e;
        } catch (Exception e) {
            throw new ExecutionException("Function execution failed: " + e.getMessage(), e);
        } finally {
            if (pooledContext != null) {
                contextPool.returnContext(pooledContext);
            } else if (ctx != null) {
                Context.exit();
            }
        }
    }

    /** Creates a null JavaScript value. */
    public Object createNullValue() {
        return null;
    }

    /** Creates a JavaScript value from a Java object. */
    public Object createValue(Object obj) {
        return obj; // Rhino handles Java objects directly
    }

    /** Creates a JavaScript object from a Map. */
    public Scriptable createObject(Map<String, Object> map) {
        RhinoPooledContext pooled = null;
        Context ctx = null;
        try {
            pooled = contextPool.borrowContext();
            ctx = pooled.getContext();
            Scriptable scope = ctx.initStandardObjects();

            Scriptable result = ctx.newObject(scope);
            for (Map.Entry<String, Object> entry : map.entrySet()) {
                result.put(entry.getKey(), result, entry.getValue());
            }
            return result;
        } catch (Exception e) {
            throw new RuntimeException("Failed to create object", e);
        } finally {
            if (pooled != null) {
                try {
                    contextPool.returnContext(pooled);
                } catch (Exception ex) {
                    // Ignore
                }
            }
        }
    }

    /** Creates a JavaScript array from an Object array. */
    public NativeArray createArray(Object[] array) {
        RhinoPooledContext pooled = null;
        Context ctx = null;
        try {
            pooled = contextPool.borrowContext();
            ctx = pooled.getContext();
            Scriptable scope = ctx.initStandardObjects();

            return (NativeArray) ctx.newArray(scope, array);
        } catch (Exception e) {
            throw new RuntimeException("Failed to create array", e);
        } finally {
            if (pooled != null) {
                try {
                    contextPool.returnContext(pooled);
                } catch (Exception ex) {
                    // Ignore
                }
            }
        }
    }

    /** Exception thrown when script execution fails. */
    public static class ExecutionException extends Exception {
        public ExecutionException(String message) {
            super(message);
        }

        public ExecutionException(String message, Throwable cause) {
            super(message, cause);
        }
    }
}

File: engine/src/sharedJvmMain/kotlin/dev/haipham22/leechtext/plugin/js/api/JsScriptEngine.kt
package dev.haipham22.leechtext.plugin.js.api

import dev.haipham22.leechtext.log.EngineLogger
import org.mozilla.javascript.Context
import org.mozilla.javascript.Function
import org.mozilla.javascript.NativeArray
import org.mozilla.javascript.Scriptable

/**
 * Script engine với pooling + resource limits (port từ JsScriptEngine.java).
 */
class JsScriptEngine private constructor(
    log: EngineLogger,
) {
    private val contextPool = RhinoContextPool(log, MAX_CONTEXTS, MAX_EXECUTION_TIME_MS, MAX_MEMORY_MB)

    fun execute(script: String): Any? {
        var pooledContext: RhinoPooledContext? = null
        var ctx: Context? = null
        try {
            pooledContext = contextPool.borrowContext()
            ctx = pooledContext.context
            ctx.languageVersion = 200
            val scope = ctx.initStandardObjects()
            return ctx.evaluateString(scope, script, "script", 1, null)
        } catch (e: Exception) {
            pooledContext?.let { contextPool.invalidateContext(it) }
            // Đặt null để finally không return lại context vừa invalidate mà exit Context
            pooledContext = null
            // invalidateContext đã Context.exit() — null luôn ctx để finally không exit lần 2
            ctx = null
            throw ExecutionException("Script execution failed: ${e.message}", e)
        } finally {
            if (pooledContext != null) {
                contextPool.returnContext(pooledContext)
            } else {
                ctx?.let { Context.exit() }
            }
        }
    }

    @Suppress("ThrowsCount") // validation chain ở trust boundary — mỗi guard ném exception có message riêng cho UI
    fun executeFunction(
        script: String,
        functionName: String,
        vararg args: Any?,
    ): Any? {
        var pooledContext: RhinoPooledContext? = null
        var ctx: Context? = null
        try {
            pooledContext = contextPool.borrowContext()
            ctx = pooledContext.context
            ctx.languageVersion = 200
            val scope = ctx.initStandardObjects()

            ctx.evaluateString(scope, script, "script", 1, null)

            val functionObj = scope[functionName, scope]
            if (functionObj !is Function) {
                throw ExecutionException("Function '$functionName' not found or not executable")
            }
            return functionObj.call(ctx, scope, scope, args)
        } catch (e: ExecutionException) {
            throw e
        } catch (e: Exception) {
            throw ExecutionException("Function execution failed: ${e.message}", e)
        } finally {
            if (pooledContext != null) {
                contextPool.returnContext(pooledContext)
            } else {
                ctx?.let { Context.exit() }
            }
        }
    }

    fun createObject(map: Map<String, Any?>): Scriptable {
        var pooled: RhinoPooledContext? = null
        return try {
            pooled = contextPool.borrowContext()
            val ctx = pooled.context
            val scope = ctx.initStandardObjects()
            val result = ctx.newObject(scope)
            for ((key, value) in map) result.put(key, result, value)
            result
        } catch (e: Exception) {
            throw IllegalStateException("Failed to create object", e)
        } finally {
            pooled?.let {
                try {
                    contextPool.returnContext(it)
                } catch (ex: Exception) {
                    // ignore
                }
            }
        }
    }

    fun createArray(array: Array<Any?>): NativeArray {
        var pooled: RhinoPooledContext? = null
        return try {
            pooled = contextPool.borrowContext()
            val ctx = pooled.context
            val scope = ctx.initStandardObjects()
            ctx.newArray(scope, array) as NativeArray
        } catch (e: Exception) {
            throw IllegalStateException("Failed to create array", e)
        } finally {
            pooled?.let {
                try {
                    contextPool.returnContext(it)
                } catch (ex: Exception) {
                    // ignore
                }
            }
        }
    }

    class ExecutionException(
        message: String,
        cause: Throwable? = null,
    ) : Exception(message, cause)

    companion object {
        private const val MAX_CONTEXTS = 5
        private const val MAX_EXECUTION_TIME_MS = 30_000
        private const val MAX_MEMORY_MB = 50

        private val LOCK = Any()
        private var engine: JsScriptEngine? = null

        @JvmStatic
        fun getInstance(log: EngineLogger): JsScriptEngine {
            if (engine == null) {
                synchronized(LOCK) {
                    if (engine == null) {
                        engine = JsScriptEngine(log)
                        log.add("Rhino JavaScript engine initialized")
                    }
                }
            }
            return engine!!
        }

        @JvmStatic
        fun resetInstance() {
            synchronized(LOCK) { engine = null }
        }
    }
}
